lib/chutney/TorNet.py

```diff
@@ -1448,35 +2086,142 @@ def hup(self):
         controllers = [n.getController() for n in self._nodes]
-        elapsed = 0.0
-        most_recent_status = [ None ] * len(controllers)
+        most_recent_bootstrap_status = [ None ] * len(controllers)
+        most_recent_desc_status = dict()
         while True:
             all_bootstrapped = True
-            most_recent_status = [ ]
+            most_recent_bootstrap_status = [ ]
+            most_recent_desc_status = dict()
             for c in controllers:
-                pct, kwd, msg = c.getLastBootstrapStatus()
-                most_recent_status.append((pct, kwd, msg))
-                if pct != 100:
+                nick = c.getNick()
+                pct, kwd, bmsg = c.getLastBootstrapStatus()
+                most_recent_bootstrap_status.append((pct, kwd, bmsg))
+
+                if pct != LocalNodeController.SUCCESS_CODE:
                     all_bootstrapped = False
 
+                desc_status = c.getNodeDirInfoStatus()
+                if desc_status:
+                    code, nodes, docs, dmsg = desc_status
+                    most_recent_desc_status[nick] = (code,
+                                                     nodes,
+                                                     docs,
+                                                     dmsg)
+                    if code != LocalNodeController.SUCCESS_CODE:
+                        all_bootstrapped = False
+
+            now = time.time()
+            elapsed = now - start
             if all_bootstrapped:
-                print("Everything bootstrapped after %s sec"%elapsed)
+                print("Everything bootstrapped after {} sec"
+                      .format(int(elapsed)))
+                self.print_bootstrap_status(controllers,
+                                            most_recent_bootstrap_status,
+                                            most_recent_desc_status,
+                                            elapsed=elapsed,
+                                            msg="Bootstrap finished")
+                # Avoid a race condition where:
+                #  - all the directory info that chutney checks is present,
+                #  - but some unchecked dir info is missing
+                #    (perhaps microdescriptors, see #33428)
+                #    or some other state or connection isn't quite ready, and
+                #  - chutney's SOCKS connection puts tor in a failing state,
+                #    which affects tor for at least 10 seconds.
+                #
+                # We have only seen this race condition in 0.3.5. The fixes to
+                # microdescriptor downloads in 0.4.0 or 0.4.1 likely resolve
+                # this issue.
+                print("Waiting {} seconds for other dir info to sync...\n"
+                      .format(int(Network.WAIT_FOR_UNCHECKED_DIR_INFO_DELAY)))
+                time.sleep(Network.WAIT_FOR_UNCHECKED_DIR_INFO_DELAY)
```

Comment:
Should we note down somewhere the places where we are sleeping where "technically" we should be awaiting some event of some sort?